src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1,10 +1,174 @@
+//! # Fingerprints
+//!
+//! This module implements change-tracking so that Cargo can know whether or
+//! not something needs to be recompiled. A Cargo `Unit` can be either "dirty"
+//! (needs to be recompiled) or "fresh" (it does not need to be recompiled).
+//! There are several mechanisms that influence a Unit's freshness:
+//!
+//! - The `Metadata` hash isolates each Unit on the filesystem by being
+//!   embedded in the filename. If something in the hash changes, then the
+//!   output files will be missing, and the Unit will be dirty (missing
+//!   outputs are considered "dirty").
+//! - The `Fingerprint` is another hash, saved to the filesystem in the
+//!   `.fingerprint` directory, that tracks information about the inputs to a
+//!   Unit. If any of the inputs changes from the last compilation, then the
+//!   Unit is considered dirty. A missing fingerprint (such as during the
+//!   first build) is also considered dirty.
+//! - Dirty propagation is done in the `JobQueue`. When a Unit is dirty, the
+//!   `JobQueue` automatically treats anything that depends on it as dirty.
+//!   Anything that relies on this is probably a bug. The fingerprint should
```

Comment:
This is definitely something we slowly realized over time, and I think at this point we should probably remove this logic in `JobQueue` perhaps? (fine to leave to a separate PR). 

It may uncover a bug or two either in the test suite or in the wild, but it's all stuff we'd have to fix eventually anyway.

Reply:
That might be a worthwhile experiment, at least to see where things break in the testsuite.